### ISSUE TITLE
Remove opensslconf.h as a public header

### ISSUE THREAD

**Issue by kaduk** [openssl/openssl#4741]
A comment on #4732 ponders whether opensslconf.h should be a public header at all.
This is potentially coupled to how many feature en/disablement knobs are exposed at build time and how those knobs interplay with our promise of API/ABI stability.
I think one could defend the claim that we currently have too many configure-visible feature en/disablement knobs, though perhaps not the claim that we should aim to have precisely zero such knobs.

**Comment by levitte** [comment]
I agree that there are knobs that don't need to be public.  I would also say that there are entire features that have no business being public (the locking mechanism is one of them).  I'm not sure we'll get rid of opensslconf.h entirely, though, like you say...

**Comment by levitte** [comment]
If we aim for 1.2.0 to have algos accessible through EVP only, there's a *lot* that can be made private...

**Comment by davidben** [comment]
It's quite common for OpenSSL consumers to use the OPENSSL_NO_* symbols to enable or disable the corresponding calling code on their end. Bindings libraries particularly like to do this. They don't have any particular requirement on one feature or another but would just like to expose whatever is available and skip exposing things that aren't available.

**Comment by dot-asm** [comment]
> It's quite common for OpenSSL consumers to use the OPENSSL_NO_* symbols to enable or disable the corresponding calling code on their end.

It's easy to let past dominate future, but it doesn't necessarily make it right. And question is not only about future, but even how to get there. Is there something one can do already *today*? We already decided not to export new low-level entry points. We can as well decide not to export EVP_newalgo and make applications look it up by its name (or nid). It wouldn't break any application and best of all it would nudge developers in desired direction already today. In other words we can aim for "no new OPNESSL_NO_THIS_N_THAT in 1.1.1 (in comparison to 1.1.0)" which transforms "there will be *no* OPENSSL_NO_THAT_N_THAT in 1.2.0".

**Comment by levitte** [comment]
> (or nid)

That implies the existence of `ERR_newalgo` macros ;-) ...

**Comment by levitte** [comment]
(delete comment...  I was talking too fast)

**Comment by levitte** [comment]
Hmm...  "no new OPENSSL_NO_THIS_OR_THAT" sounds like a good idea, but executing that might be trickier...  or well, it's really "just" a question of tweaking Configure...  let's hope we don't make it too confusing.

**Comment by dot-asm** [comment]
> "no new OPENSSL_NO_THIS_OR_THAT" [is] a question of tweaking Configure...

Just to be clear. Suggestions is *not* to remove opt-in/-out mechanism, but to make application adapt at run time. I mean having application rely on OPENSSL_NO_THIS_N_THAT is compile-time decision, and having application query availability is run-time one. So it's not like opensslconf.h would disappear, it just won't be public. And as *intermediate* state for 1.1.x, it would be viewed as frozen.

Possible alternative (to looking up by name) is to have all EVP_algo exported, all there are that is, but those that are opted out would return NULL.

**Comment by dot-asm** [comment]
> as intermediate state for 1.1.x, it would be viewed as frozen.

I naturally meant public opensslconf.h would be viewed as frozen. As intermediate state there would be internal/opensslconf.h that would include public. Though of course keeping track of disables in 1.1.0 and post-1.1.0 is something that would have to be done in Configure. I mean 1.1.0 disables would go to public opensslconf.h and post-1.1.0 - to private opensslconf.h...

**Comment by levitte** [comment]
Yes, understood, all we need to do is distinguish between the two.

**Comment by dot-asm** [comment]
It looks like the only remaining question is can we agree that it's desirable goal to pursue. Well, three seem to agree, but we still need more opinions...

**Comment by levitte** [comment]
Let me just say that..  "there be dragons".  I just took a stab at doing this with something as mundane as SM4, and when I realised that we might need a specialised variant `IMPLEMENT_BLOCK_CIPHER` just to implement EVP_sm4_ecb() et al for the disabled case, I threw what I'd done away.  I thought it would be a fairly simple hack :-/

So either I went down the wrong route and need to rethink this, or we have to understand that moving things to be checkable in run-time rather than compile-time is a big change.

This has me think that entirely removing opensslconf.h from public view is probably not a good move.  However, there are certainly things in there that can be hidden away (like the definition of BN_LLONG, or of OPENSSL_UNISTD, or ...  basically all stuff that's just internal configuration and not declarations of available features).

**Comment by dot-asm** [comment]
> Let me just say that.. "there be dragons". ... moving things to be checkable in run-time rather than compile-time is a big change.

Yes. And? Life is, has been and will be full of compromises, i.e. some things will be worth, but others will be better. It's question about priorities.

> I thought it would be a fairly simple hack :-/

??? As proof of concept it's sufficient to simply remove EVP_sm4 lines from util/libcrypto.num and include/openssl/evp.h. `make test` does pass and produces expected result without and with no-sm4. "Expected result without no-sm4" means that it is exercised in evp_test and you can use it as argument to `enc`. Do note "as proof of concept", as I'm not saying that this absolutely sufficient in general case...

**Comment by levitte** [comment]
I was going along the route of keeping `EVP_sm4` (in all its variants) and having them return NULL when SM4 is disabled... Of course it's much easier to remove them entirely. Are our users ready for that, especially for having them all removed in 1.2.0? (on the other hand, it's not too hard to have them in a compatibility layer that just calls the general algorithm finder with appropriate nid symbols)

**Comment by dot-asm** [comment]
> I was going along the route of keeping EVP_sm4 (in all its variants) and having them return NULL when SM4 is disabled...

Even that should be as trivial as #ifndef OPENSSL_NO_SM4 whole crypto/evp/e_sm4.c and adding stubs in #else... No? Keep in mind that opensslconf.h doesn't have to disappear, suggestion is make it private.

> Are our users ready for that, especially for having them all removed in 1.2.0?

As already said, it's easy to let past dominate future, but it doesn't necessarily make it right. And we do have opportunity to define the future vector *now*. In sense that new applications that are written today can be directed toward run-time detection. I argue that in long run it *is* better, run-time detection, for all involved parties.

**Comment by dot-asm** [comment]
> having them all removed in 1.2.0

Special note about "all". So far suggestion was to make *options* to be run-time detectable. I.e. those that can be enabled or disabled at build time. Underlying rationale is that application provider doesn't have to be and even likely is not same party as library provider... But one can arguably extend it all algorithms...

**Comment by t-j-h** [comment]
Having both compile time and runtime options is important. Removing or reducing compile time feature removal wouldn't be a choice I would be positive on - it is widely used.

**Comment by noloader** [comment]
> Remove opensslconf.h as a public header

Please, no.

People have been told to check for features by including `<opensslconf.h>` and then testing a macro. For example, [How to know if my openssl supports sslv3](https://stackoverflow.com/q/30591363/608639), [Building ruby with rbenv and ruby-build fails with undefined symbol: SSLv2_method](https://stackoverflow.com/q/36286437/608639), [Openssl SSL_CTX_new(SSLv3_method()) returns NULL](https://stackoverflow.com/q/25282939/608639), etc.

If you remove the workflow, then please ensure the previosu macros are still available.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
